### PR TITLE
Write SMS package

### PR DESCRIPTION
**Pull request by gmkv**
This change introduces functionality to send SMS messages.
Using the package sms, you may send SMS messages using
```
messenger := sms.NewSender(sms.NewSyntaxValidator(), sms.NewSMSGateway())
m := sender.NewForSending(...)
err := sender.SendSMS(m SMS)
```
SendSMS internally composes a message using the supplied parameters,
validates the message, and sends it off to a SMS gateway.
If an error occurs, it will be reported appropriately.

### CODE REVIEW

File: smsApp/sms/sender.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+	// Check if SMS body is too large to fit in one sms
+	if len(m.Body) > MaxSMSLength {
+		longText := m.Split()
+
+		// Recursive call to send each SMS. 👃_______
+		for i := range longText {
+			err := s.SendSMS(longText[i])
+			if err != nil {
+				return err
```
Comment (mgenov) [inline smsApp/sms/sender.go]:
```golang
return fmt.Errorf("unable to send sms due: %v")
```

File: smsApp/sms/sender.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+			}
+
+			/*
+				err := s.v.ValidateSMS(longText[i])
```
Comment (mgenov) [inline smsApp/sms/sender.go]: Commented code need to be deleted.

File: smsApp/sms/sms.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+type SMS struct {
+	To   string
+	From string
+	Body string
+}
+
+// New returns a new SMS instance
+func NewSMS() SMS {
```
Comment (mgenov) [inline smsApp/sms/sms.go]: sms.NewSMS() is duplicating the SMS in name, which is smelling. 

message := sms.New()

File: smsApp/sms/sms.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	return SMS{}
+}
+
+// Compose returns an SMS initialised using the supplied parameters
+func Compose(to string, from string, body string) SMS {
```
Comment (mgenov) [inline smsApp/sms/sms.go]: NewForSending()

File: smsApp/smsMain.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package main
```
Comment (mgenov) [inline smsApp/smsMain.go]: It is better to be placed in:
cmsApp/cmd/app/main.go 

as in Go in most cases binaries are placed in the cmd folder.

File: smsApp/sms/sms.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		result[0] = m
+		return result
+	}
+
+	var result []SMS
+	var chunk string
+	length := len(m.Body)
+
+	// Split the long message on every 140 symbols in chunks.
+	for i := 0; i < length; i += MaxSMSLength {
```
Comment (mgenov) [inline smsApp/sms/sms.go]: The min function in the loop looks really strange and hard to understand (At least for me). 

Here is a snippet of split function which uses runes:

```golang
func SplitSubN(s string, n int) []string {
    sub := ""
    subs := []string{}

    runes := bytes.Runes([]byte(s))
    l := len(runes)
    for i, r := range runes {
        sub = sub + string(r)
        if (i + 1) % n == 0 {
            subs = append(subs, sub)
            sub = ""
        } else if (i + 1) == l {
            subs = append(subs, sub)
        }
    }

    return subs
}
```

File: smsApp/sms/validator_test.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
+package sms_test
+
+import (
+	"testing"
+
+	"github.com/clouway-incubator/georgi_minkov_go_testing/smsApp/sms"
+)
+
+var ValidValues = []struct {
```
Comment (mgenov) [inline smsApp/sms/validator_test.go]: It should be in lower case as it's not an Exported type: 

var cases = []struct {
}

File: smsApp/sms/sms_test.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	return true
+}
+
+func TestCompose(t *testing.T) {
+
+	for _, row := range composeValues {
+
+		got := sms.Compose(row.passedArgs.to, row.passedArgs.from, row.passedArgs.body)
+
+		if !equal(got, row.expected) {
```
Comment (mgenov) [inline smsApp/sms/sms_test.go]: Take a look at: https://golang.org/pkg/reflect/#DeepEqual

This is going to remove most of the code here.

File: smsApp/sms/sms_test.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	if a.Body != b.Body {
+		return false
+	}
+
+	return true
+}
+
+func TestCompose(t *testing.T) {
```
Comment (mgenov) [inline smsApp/sms/sms_test.go]: This tests is not required as you are duplicating it with the validation. e.g if SMS is not initialized then the validation will fail which sounds like contract.

**Comment by gmkv** [comment]
I restructured the files and folders a bit, removed unnecessary commented code, changed function and variables' names, changed sms.Split. I changed the unit test in sms_test.go to verify Split func, rather than the NewForSending func.

File: sms-app/sms/sender.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
+package sms
+
+import "fmt"
+
+type Sender struct {
+	v Validator
```
Comment (mgenov) [inline sms-app/sms/sender.go]: validator
and 
gateway 

v and g are too short .

File: sms-app/sms/sender_test.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+		err := sender.SendSMS(row)
+
+		if err != nil {
+			t.Errorf("Error received from sender.SendSMS(%v) = %v\n", row, err)
+		}
+	}
+}
+
+func TestSendSMSWithCompose(t *testing.T) {
```
Comment (mgenov) [inline sms-app/sms/sender_test.go]: Not able to see cases when validation fails or send fails.

Comment (gmkv) [reply sms-app/sms/sender_test.go]: How is a good way to go about including these cases?

I can think of defining behaviour to the mocked objects in a certain order with gomock.InOrder(...), so the mocked objects will return nil when expected, and will error out when expected.
For example:
```
gomock.InOrder(
		mockGateway.EXPECT().SendToGateway(gomock.Any()).Return(nil).Times(8),
		mockGateway.EXPECT().SendToGateway(gomock.Any()).Return(sms.ErrGatewayExternalValidationFail).Times(1),
	)
```
And also to restructure the cases adding what's also going to be expected to happen, similar to my test in validator_test.go:
```
var sendSMSCases = []struct {
	passed   sms.SMS
	expected error
}{...}
```

Comment (mgenov) [reply sms-app/sms/sender_test.go]: Don't use table driven tests for this scenario as they relay on mock objects and things are a little bit different then simple state tests. 

About inOrder you can create specific test that ensures that validation is executed before testing:
TestValidationAndSendingAreExecutedInOrder 

e.g every test should test specific behaviour and should not mess most of them. 

Overall: 
TestValidationFails
TestSmsIsPassedToValidator (expect param passed to validator)
TestSendFails
TestSmsIsPassedForSending (expect param passed to the Gateway)
TestValidationAndSendingAreExecutedInOrder

Comment (gmkv) [reply sms-app/sms/sender_test.go]: Would you please take a look at my commit when you have the time and let me know if I'm on the right track? I'm lost here on figuring out TestValidationAndSendingAreExecutedInOrder.

**Comment by gmkv** [comment]
I have changed the tests and squashed the commits.

**Pull request merged**